src/uinput.rs

```diff
@@ -303,10 +303,12 @@ impl VirtualDevice {
     ///
     /// The returned event allows the user to allocate and set the effect ID as well as access the
     /// effect data.
-    pub fn process_ff_upload(&mut self, event: UInputEvent) -> Result<FFUploadEvent, Error> {
-        if event.code() != UInputCode::UI_FF_UPLOAD {
-            return Err(Error::InvalidEvent);
-        }
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `event.code()` is not `UI_FF_UPLOAD`.
+    pub fn process_ff_upload(&mut self, event: UInputEvent) -> io::Result<FFUploadEvent> {
+        assert_eq!(event.code(), UInputCode::UI_FF_UPLOAD);
```

Comment:
Hmm... if we didn't want to introduce an explicit panic here, we could just return `io::ErrorKind::Other` with a message.

Reply:
Yeah, and that's what I did at first, but I feel like this is the kinda thing panics are for - if you passed a value that doesn't make sense to this function, you've made a logic mistake in your program; it's not an error you can pass back up to the user or something.

Reply:
At that point though, why even have a mandatory parameter that must be a specific value? Maybe we should just remove the param entirely.

Reply:
the other data that comes in on the event is used though, right?

Reply:
It is, yeah - the value and maybe-but-probably-not the timestamp are relevant.

Reply:
Hmm... I'm generally more in favor of allowing callers of a library to decide whether or not to panic on something like this, seeing as some folks might be trying to use evdev in embedded apps like kiosks and the like. Would it hurt to return `std::io::Error` with `ErrorKind::Other`?